pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -669,7 +669,7 @@ func (a *Actuator) ensureAnnotation(ctx context.Context, machine *machinev1beta1
 	if err := a.client.Update(ctx, machine); err != nil {
 		return gherrors.Wrap(err, "failed to update machine annotation")
 	}
```

Comment:
Could add a comment here to the effect that "The update to the annotations will trigger a requeue once it is observed, but we must return RequeueAfterError so the controller knows the operation is not complete. Use a delay so that we don't requeue _before_ the annotation change is observed."

Reply:
> "we must return RequeueAfterError so the controller knows the operation is not complete"

Is that some state management logic specific to this controller? Have we added meaning to returning without error, that some imperative action is "complete"?

Reading the BZ, it sounds like that's the case. "we have to return RequeueAfterError{} to prevent the controller deciding we have completed the Create phase prematurely."

Adding some significant explanation here would be useful. This delay sounds like a workaround for a design flaw that may be worth re-considering.

Reply:
Yes, if the actuator doesn't return an error from Create() then the controller decides that creation is complete:

https://github.com/openshift/cluster-api-provider-baremetal/blob/master/vendor/github.com/openshift/machine-api-operator/pkg/controller/machine/controller.go#L365-L377

There's no error type you can return that doesn't requeue. If you return requeueAfterError then we'll requeue after, and if you return any other error then the error will propagate (which obviously causes an immediate requeue as well, in addition to tracebacks in the logs).

Reply:
OH RIGHT the actuator pattern. Yep, it's ripe with these issues.

Ok, but is there a negative consequence to it thinking creation is complete? A repeated log entry saying it was created could be annoying, but it'll use the return value of `Exists()` to decide if it should call Create again. It looks like without an error from `Create()`, the controller requeues after 30 seconds. Presumably that's a sufficient delay.

I don't mean to block this, and there's no need to wait for me to comment further on this issue unless you want to keep discussing it. The proposed change would be fine, but I wanted to at least spark a little discussion of alternatives.

In general I support subverting the actuator pattern as needed to make a robust controller. If there's an opportunity to get rid of the race condition by changing how we use the actuator pattern (for example making sure `Exists()` keeps returning `false` until all of the work in `Create()` is done, or moving the annotation assurance to the `Update()` method), I think that's worth considering.

Reply:
> OH RIGHT the actuator pattern.

:grinning:

> Ok, but is there a negative consequence to it thinking creation is complete? A repeated log entry saying it was created could be annoying, but it'll use the return value of `Exists()` to decide if it should call Create again. It looks like without an error from `Create()`, the controller requeues after 30 seconds. Presumably that's a sufficient delay.

Yeah, you're right (in fact 30s is the delay we're returning), I forgot how it worked didn't look close enough. So returning nil to the controller and what we're doing here are equivalent except for that log message. So maybe just the last sentence of my suggested comment is sufficient.

Another reason we're returning an error here (which we could easily change) is that we're writing to the Machine here and the next call in Create(), `clearInsufficientResourcesError()` writes to the Status - so the latter will fail and log an error 100% of the time if we call it immediately after, because the object will be out of date. We're currently checking only the error to indicate whether there was a change, and we'd need to add another return value otherwise.